File: src/core/companion/protocol.ts
import {
  COMPANION_MAX_MESSAGE_BYTES,
  COMPANION_PROTOCOL_VERSION,
  type CompanionAuthBundle,
  type CompanionFailure,
  type CompanionFallback,
  type CompanionHealth,
  type CompanionJobProgress,
  type CompanionOutputReceipt,
  type YtDlpMediaSummary,
} from "./types";

export interface CompanionEnvelope<T = unknown> {
  protocolVersion: 1;
  requestId: string;
  type: string;
  payload: T;
}

export type CompanionRequest =
  | CompanionEnvelope<{ browserTarget: "brave" | "chrome" | "chromium" | "unknown"; extensionVersion: string }>
  | CompanionEnvelope<{ pageUrl: string; auth: CompanionAuthBundle }>
  | CompanionEnvelope<{
      jobId: string;
      probeToken: string;
      selectionKey: string;
      auth: CompanionAuthBundle;
    }>
  | CompanionEnvelope<{
      jobId: string;
      probeToken: string;
      selectionKey: string;
      clip: { startMs: number; endMs: number; mode: "fast" | "exact" };
      allowFullDownloadFallback: boolean;
      auth: CompanionAuthBundle;
    }>
  | CompanionEnvelope<{ jobId: string }>
  | CompanionEnvelope<{ outputToken: string }>
  | CompanionEnvelope<Record<string, never>>;

export type CompanionEvent =
  | CompanionEnvelope<CompanionHealth>
  | CompanionEnvelope<YtDlpMediaSummary>
  | CompanionEnvelope<CompanionFailure>
  | CompanionEnvelope<{ jobId: string; position: number }>
  | CompanionEnvelope<CompanionJobProgress>
  | CompanionEnvelope<CompanionFallback>
  | CompanionEnvelope<CompanionOutputReceipt>
  | CompanionEnvelope<{ jobId: string }>
  | CompanionEnvelope<{ stage: string; percentage?: number; detail?: string }>;

const REQUEST_TYPES = new Set([
  "hello",
  "probe",
  "start_download",
  "start_clip",
  "cancel_job",
  "reveal_output",
  "open_output",
  "install_tools",
  "update_tools",
]);

const EVENT_TYPES = new Set([
  "hello_result",
  "probe_result",
  "auth_required",
  "job_queued",
  "job_progress",
  "fallback_required",
  "job_completed",
  "job_failed",
  "job_cancelled",
  "tool_progress",
]);

const ERROR_CODES = new Set([
  "COMPANION_NOT_INSTALLED",
  "COMPANION_INCOMPATIBLE",
  "TOOLS_MISSING",
  "TOOLS_INCOMPATIBLE",
  "PROTOCOL_MISMATCH",
  "URL_UNSUPPORTED",
  "PLAYLIST_UNSUPPORTED",
  "LIVE_UNSUPPORTED",
  "DRM_UNSUPPORTED",
  "AUTH_REQUIRED",
  "AUTH_SCOPE_INSUFFICIENT",
  "COOKIE_PERMISSION_DENIED",
  "FORMAT_UNAVAILABLE",
  "SECTION_UNSUPPORTED",
  "EXACT_CLIP_UNSUPPORTED",
  "OUTPUT_EXISTS",
  "DISK_FULL",
  "JOB_INTERRUPTED",
  "CANCELLED",
  "INVALID_REQUEST",
  "INTERNAL_ERROR",
]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function serializedSize(value: unknown): number {
  return new TextEncoder().encode(JSON.stringify(value)).byteLength;
}

function onlyKeys(record: Record<string, unknown>, allowed: readonly string[]): boolean {
  const set = new Set(allowed);
  return Object.keys(record).every((key) => set.has(key));
}

function boundedString(value: unknown, max = 4096): value is string {
  return typeof value === "string" && value.length > 0 && value.length <= max;
}

function optionalBoundedString(value: unknown, max = 4096): boolean {
  return value === undefined || boundedString(value, max);
}

function nonNegativeNumber(value: unknown): boolean {
  return typeof value === "number" && Number.isFinite(value) && value >= 0;
}

function optionalNonNegativeNumber(value: unknown): boolean {
  return value === undefined || nonNegativeNumber(value);
}

function integer(value: unknown, minimum = 0): boolean {
  return Number.isSafeInteger(value) && (value as number) >= minimum;
}

function httpUrl(value: unknown): value is string {
  if (!boundedString(value, 8192)) return false;
  try {
    const protocol = new URL(value).protocol;
    return protocol === "http:" || protocol === "https:";
  } catch {
    return false;
  }
}

function validAuth(value: unknown): boolean {
  if (!isRecord(value) || !boundedString(value.mode, 32)) return false;
  if (value.mode === "anonymous") return onlyKeys(value, ["mode"]);
  if (value.mode === "brave-profile") {
    return onlyKeys(value, ["mode", "profileId"]) && boundedString(value.profileId, 128);
  }
  if (value.mode !== "current-tab") return false;
  if (!onlyKeys(value, ["mode", "pageUrl", "referer", "userAgent", "cookieStoreId", "incognito", "cookies"])) return false;
  if (!httpUrl(value.pageUrl) || !httpUrl(value.referer) || !boundedString(value.userAgent, 1024) ||
      !boundedString(value.cookieStoreId, 128) || typeof value.incognito !== "boolean" ||
      !Array.isArray(value.cookies) || value.cookies.length > 2048) return false;
  return value.cookies.every((cookie) => {
    if (!isRecord(cookie) || !onlyKeys(cookie, ["name", "value", "domain", "path", "secure", "httpOnly", "sameSite", "expirationDate", "hostOnly", "session"])) return false;
    return typeof cookie.name === "string" && cookie.name.length <= 1024 &&
      typeof cookie.value === "string" && cookie.value.length <= 8192 &&
      boundedString(cookie.domain, 1024) && boundedString(cookie.path, 2048) &&
      typeof cookie.secure === "boolean" && typeof cookie.httpOnly === "boolean" &&
      ["no_restriction", "lax", "strict", "unspecified"].includes(String(cookie.sameSite)) &&
      (cookie.expirationDate === undefined || nonNegativeNumber(cookie.expirationDate)) &&
      typeof cookie.hostOnly === "boolean" && typeof cookie.session === "boolean";
  });
}

function validSelection(value: unknown): boolean {
  if (!isRecord(value) || !onlyKeys(value, ["kind", "key", "label", "estimatedBytes", "expectedContainer", "videoFormatId", "audioFormatId"])) return false;
  if ((value.kind !== "preset" && value.kind !== "formats") || !boundedString(value.key, 128) || !boundedString(value.label, 512)) return false;
  if (!optionalNonNegativeNumber(value.estimatedBytes) || !optionalBoundedString(value.expectedContainer, 32) ||
      !optionalBoundedString(value.videoFormatId, 128) || !optionalBoundedString(value.audioFormatId, 128)) return false;
  if (value.kind === "preset" && !["best", "best-mp4", "up-to-1080p", "up-to-720p", "audio-only"].includes(String(value.key))) return false;
  return true;
}

function validSummary(value: unknown): boolean {
  if (!isRecord(value) || !onlyKeys(value, ["extractorKey", "mediaId", "webpageUrl", "title", "durationMs", "thumbnailUrl", "uploader", "isLive", "isDrm", "selections", "probeToken", "probedAt"])) return false;
  return boundedString(value.extractorKey, 128) && boundedString(value.mediaId, 512) &&
    httpUrl(value.webpageUrl) && boundedString(value.title, 4096) &&
    (value.durationMs === undefined || integer(value.durationMs)) &&
    (value.thumbnailUrl === undefined || httpUrl(value.thumbnailUrl)) &&
    optionalBoundedString(value.uploader, 1024) && typeof value.isLive === "boolean" &&
    (value.isDrm === undefined || typeof value.isDrm === "boolean") &&
    Array.isArray(value.selections) && value.selections.length <= 128 &&
    value.selections.every(validSelection) && boundedString(value.probeToken, 512) && integer(value.probedAt);
}

function validFailure(value: unknown): boolean {
  return isRecord(value) && onlyKeys(value, ["jobId", "code", "message", "recoverable"]) &&
    optionalBoundedString(value.jobId, 128) && ERROR_CODES.has(String(value.code)) &&
    boundedString(value.message, 4096) && typeof value.recoverable === "boolean";
}

function validHealth(value: unknown): boolean {
  if (!isRecord(value) || !onlyKeys(value, ["protocolVersion", "companionVersion", "browserTarget", "platform", "ytDlpVersion", "ffmpegVersion", "ffprobeVersion", "jsRuntime", "healthy", "issues", "capabilities", "braveProfiles"])) return false;
  if (value.protocolVersion !== 1 || !boundedString(value.companionVersion, 64) ||
      !["brave", "chrome", "chromium", "unknown"].includes(String(value.browserTarget)) ||
      !["macos", "windows", "linux"].includes(String(value.platform)) ||
      !optionalBoundedString(value.ytDlpVersion, 128) ||
      !optionalBoundedString(value.ffmpegVersion, 128) ||
      !optionalBoundedString(value.ffprobeVersion, 128) ||
      typeof value.healthy !== "boolean" || !Array.isArray(value.issues) || value.issues.length > 64 ||
      !isRecord(value.capabilities)) return false;
  if (value.jsRuntime !== undefined &&
      (!isRecord(value.jsRuntime) || !onlyKeys(value.jsRuntime, ["name", "version"]) ||
       !boundedString(value.jsRuntime.name, 64) || !boundedString(value.jsRuntime.version, 128))) return false;
  const capabilities = value.capabilities;
  if (!onlyKeys(capabilities, ["probe", "download", "sectionDownload", "exactClip", "currentTabCookies", "braveProfileCookies", "revealOutput"]) ||
      !Object.values(capabilities).every((entry) => typeof entry === "boolean")) return false;
  if (!value.issues.every(validFailure)) return false;
  return value.braveProfiles === undefined || (Array.isArray(value.braveProfiles) && value.braveProfiles.length <= 128 && value.braveProfiles.every((profile) =>
    isRecord(profile) && onlyKeys(profile, ["id", "name"]) && boundedString(profile.id, 128) && boundedString(profile.name, 512)));
}

function validRequestPayload(type: string, payload: Record<string, unknown>): boolean {
  switch (type) {
    case "hello":
      return onlyKeys(payload, ["browserTarget", "extensionVersion"]) &&
        ["brave", "chrome", "chromium", "unknown"].includes(String(payload.browserTarget)) &&
        boundedString(payload.extensionVersion, 64);
    case "probe":
      return onlyKeys(payload, ["pageUrl", "auth"]) && httpUrl(payload.pageUrl) && validAuth(payload.auth);
    case "start_download":
      return onlyKeys(payload, ["jobId", "probeToken", "selectionKey", "auth"]) &&
        boundedString(payload.jobId, 128) && boundedString(payload.probeToken, 512) &&
        boundedString(payload.selectionKey, 128) && validAuth(payload.auth);
    case "start_clip": {
      if (!onlyKeys(payload, ["jobId", "probeToken", "selectionKey", "clip", "allowFullDownloadFallback", "auth"]) ||
          !boundedString(payload.jobId, 128) || !boundedString(payload.probeToken, 512) ||
          !boundedString(payload.selectionKey, 128) || typeof payload.allowFullDownloadFallback !== "boolean" ||
          !validAuth(payload.auth) || !isRecord(payload.clip)) return false;
      const clip = payload.clip;
      return onlyKeys(clip, ["startMs", "endMs", "mode"]) && integer(clip.startMs) && integer(clip.endMs, 1) &&
        (clip.endMs as number) > (clip.startMs as number) && (clip.mode === "fast" || clip.mode === "exact");
    }
    case "cancel_job":
      return onlyKeys(payload, ["jobId"]) && boundedString(payload.jobId, 128);
    case "reveal_output":
    case "open_output":
      return onlyKeys(payload, ["outputToken"]) && boundedString(payload.outputToken, 512);
    case "install_tools":
    case "update_tools":
      return onlyKeys(payload, []);
    default:
      return false;
  }
}

function validEventPayload(type: string, payload: Record<string, unknown>): boolean {
  switch (type) {
    case "hello_result": return validHealth(payload);
    case "probe_result": return validSummary(payload);
    case "auth_required":
    case "job_failed": return validFailure(payload);
    case "job_queued": return onlyKeys(payload, ["jobId", "position"]) && boundedString(payload.jobId, 128) && integer(payload.position);
    case "job_progress":
      return onlyKeys(payload, ["jobId", "stage", "downloadedBytes", "totalBytes", "percentage", "speedBytesPerSecond", "etaSeconds", "mediaRole", "detail"]) &&
        boundedString(payload.jobId, 128) && ["planning", "downloading", "merging", "processing", "saving", "completed"].includes(String(payload.stage)) &&
        optionalNonNegativeNumber(payload.downloadedBytes) && optionalNonNegativeNumber(payload.totalBytes) &&
        (payload.percentage === undefined || (nonNegativeNumber(payload.percentage) && (payload.percentage as number) <= 100)) &&
        optionalNonNegativeNumber(payload.speedBytesPerSecond) && optionalNonNegativeNumber(payload.etaSeconds) &&
        (payload.mediaRole === undefined || ["audio", "video", "combined"].includes(String(payload.mediaRole))) &&
        optionalBoundedString(payload.detail, 4096);
    case "fallback_required":
      return onlyKeys(payload, ["jobId", "reason", "estimatedBytes"]) && boundedString(payload.jobId, 128) &&
        ["section-unsupported", "section-invalid", "exact-validation-failed"].includes(String(payload.reason)) &&
        optionalNonNegativeNumber(payload.estimatedBytes);
    case "job_completed":
      return onlyKeys(payload, ["jobId", "outputToken", "filename", "finalPath", "byteSize", "container", "durationMs", "extractorKey", "mediaId", "accuracy", "actualStartMs", "actualDurationMs"]) &&
        boundedString(payload.jobId, 128) && boundedString(payload.outputToken, 512) && boundedString(payload.filename, 1024) &&
        boundedString(payload.finalPath, 8192) && integer(payload.byteSize) && boundedString(payload.container, 32) &&
        optionalNonNegativeNumber(payload.durationMs) && boundedString(payload.extractorKey, 128) && boundedString(payload.mediaId, 512) &&
        (payload.accuracy === undefined || payload.accuracy === "keyframe-aligned" || payload.accuracy === "exact") &&
        optionalNonNegativeNumber(payload.actualStartMs) && optionalNonNegativeNumber(payload.actualDurationMs);
    case "job_cancelled":
      return onlyKeys(payload, ["jobId"]) && boundedString(payload.jobId, 128);
    case "tool_progress":
      return onlyKeys(payload, ["stage", "percentage", "detail"]) && boundedString(payload.stage, 128) &&
        (payload.percentage === undefined || (nonNegativeNumber(payload.percentage) && (payload.percentage as number) <= 100)) &&
        optionalBoundedString(payload.detail, 4096);
    default:
      return false;
  }
}

export function isCompanionEnvelope(value: unknown): value is CompanionEnvelope {
  if (!isRecord(value)) return false;
  return (
    value.protocolVersion === COMPANION_PROTOCOL_VERSION &&
    typeof value.requestId === "string" &&
    value.requestId.length > 0 &&
    value.requestId.length <= 128 &&
    typeof value.type === "string" &&
    value.type.length > 0 &&
    isRecord(value.payload) &&
    serializedSize(value) <= COMPANION_MAX_MESSAGE_BYTES
  );
}

export function validateCompanionRequest(value: unknown): value is CompanionRequest {
  return isCompanionEnvelope(value) && REQUEST_TYPES.has(value.type) &&
    validRequestPayload(value.type, value.payload as Record<string, unknown>);
}

export function validateCompanionEvent(value: unknown): value is CompanionEvent {
  return isCompanionEnvelope(value) && EVENT_TYPES.has(value.type) &&
    validEventPayload(value.type, value.payload as Record<string, unknown>);
}

export function createCompanionEnvelope<T>(
  type: string,
  requestId: string,
  payload: T,
): CompanionEnvelope<T> {
  const envelope: CompanionEnvelope<T> = {
    protocolVersion: COMPANION_PROTOCOL_VERSION,
    requestId,
    type,
    payload,
  };
  if (!isCompanionEnvelope(envelope)) throw new TypeError("Invalid companion message envelope");
  return envelope;
}

File: companion/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::fmt;

pub const PROTOCOL_VERSION: u32 = 1;
pub const MAX_MESSAGE_BYTES: usize = 256 * 1024;

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Envelope {
    pub protocol_version: u32,
    pub request_id: String,
    #[serde(rename = "type")]
    pub message_type: String,
    #[serde(default)]
    pub payload: Value,
}

impl Envelope {
    pub fn response<T: Serialize>(
        request_id: impl Into<String>,
        message_type: impl Into<String>,
        payload: T,
    ) -> Result<Self, HostError> {
        Ok(Self {
            protocol_version: PROTOCOL_VERSION,
            request_id: request_id.into(),
            message_type: message_type.into(),
            payload: serde_json::to_value(payload).map_err(HostError::from)?,
        })
    }

    pub fn parse_payload<T: for<'de> Deserialize<'de>>(&self) -> Result<T, HostError> {
        serde_json::from_value(self.payload.clone())
            .map_err(|_| HostError::new(ErrorCode::InvalidRequest, "Invalid message payload"))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorCode {
    InvalidRequest,
    ProtocolMismatch,
    UrlUnsupported,
    PlaylistUnsupported,
    LiveUnsupported,
    DrmUnsupported,
    AuthRequired,
    AuthScopeInsufficient,
    FormatUnavailable,
    ProbeExpired,
    ToolsMissing,
    ToolsIncompatible,
    /// Internal distinction used for recovery routing; serialized as the
    /// stable public TOOLS_INCOMPATIBLE code.
    #[serde(rename = "TOOLS_INCOMPATIBLE")]
    ToolsOutdated,
    ToolsInstallUnavailable,
    SectionUnsupported,
    ExactClipUnsupported,
    DiskFull,
    JobInterrupted,
    Cancelled,
    OutputUnavailable,
    Internal,
}

impl ErrorCode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::InvalidRequest => "INVALID_REQUEST",
            Self::ProtocolMismatch => "PROTOCOL_MISMATCH",
            Self::UrlUnsupported => "URL_UNSUPPORTED",
            Self::PlaylistUnsupported => "PLAYLIST_UNSUPPORTED",
            Self::LiveUnsupported => "LIVE_UNSUPPORTED",
            Self::DrmUnsupported => "DRM_UNSUPPORTED",
            Self::AuthRequired => "AUTH_REQUIRED",
            Self::AuthScopeInsufficient => "AUTH_SCOPE_INSUFFICIENT",
            Self::FormatUnavailable => "FORMAT_UNAVAILABLE",
            Self::ProbeExpired => "FORMAT_UNAVAILABLE",
            Self::ToolsMissing => "TOOLS_MISSING",
            Self::ToolsIncompatible => "TOOLS_INCOMPATIBLE",
            Self::ToolsOutdated => "TOOLS_INCOMPATIBLE",
            Self::ToolsInstallUnavailable => "TOOLS_MISSING",
            Self::SectionUnsupported => "SECTION_UNSUPPORTED",
            Self::ExactClipUnsupported => "EXACT_CLIP_UNSUPPORTED",
            Self::DiskFull => "DISK_FULL",
            Self::JobInterrupted => "JOB_INTERRUPTED",
            Self::Cancelled => "CANCELLED",
            Self::OutputUnavailable => "INTERNAL_ERROR",
            Self::Internal => "INTERNAL_ERROR",
        }
    }
}

#[derive(Debug)]
pub struct HostError {
    pub code: ErrorCode,
    detail: String,
}

impl HostError {
    pub fn new(code: ErrorCode, detail: impl Into<String>) -> Self {
        Self {
            code,
            detail: detail.into(),
        }
    }

    pub fn safe_message(&self) -> String {
        crate::security::redact_diagnostic(&self.detail)
    }
}

impl fmt::Display for HostError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}: {}", self.code.as_str(), self.safe_message())
    }
}

impl std::error::Error for HostError {}

impl From<std::io::Error> for HostError {
    fn from(error: std::io::Error) -> Self {
        let code = if error.raw_os_error() == Some(28) {
            ErrorCode::DiskFull
        } else {
            ErrorCode::Internal
        };
        Self::new(code, error.to_string())
    }
}

impl From<serde_json::Error> for HostError {
    fn from(error: serde_json::Error) -> Self {
        Self::new(ErrorCode::InvalidRequest, error.to_string())
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ErrorPayload {
    pub code: String,
    pub message: String,
    pub recoverable: bool,
}

impl From<&HostError> for ErrorPayload {
    fn from(error: &HostError) -> Self {
        Self {
            code: error.code.as_str().to_owned(),
            message: error.safe_message(),
            recoverable: !matches!(error.code, ErrorCode::InvalidRequest),
        }
    }
}

#[derive(Debug, Clone, Copy, Default, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum BrowserTarget {
    Brave,
    Chrome,
    Chromium,
    #[default]
    Unknown,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct HelloRequest {
    #[serde(default)]
    pub browser_target: BrowserTarget,
    pub extension_version: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CapabilitySet {
    pub probe: bool,
    pub download: bool,
    pub section_download: bool,
    pub exact_clip: bool,
    pub current_tab_cookies: bool,
    pub brave_profile_cookies: bool,
    pub reveal_output: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HealthIssue {
    pub code: String,
    pub message: String,
    pub recoverable: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JsRuntime {
    pub name: String,
    pub version: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BraveProfile {
    pub id: String,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct InstalledReleaseInfo {
    pub release_version: String,
    pub extension_version: String,
    pub companion_version: String,
    pub tool_release_id: String,
    pub installed_at: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HelloResult {
    pub protocol_version: u32,
    pub companion_version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub installed_release: Option<InstalledReleaseInfo>,
    pub browser_target: BrowserTarget,
    pub platform: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub yt_dlp_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ffmpeg_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ffprobe_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub js_runtime: Option<JsRuntime>,
    pub healthy: bool,
    pub issues: Vec<HealthIssue>,
    pub capabilities: CapabilitySet,
    pub brave_profiles: Vec<BraveProfile>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct ProbeRequest {
    pub page_url: String,
    pub auth: AuthRequest,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(tag = "mode", rename_all = "kebab-case")]
#[serde(rename_all_fields = "camelCase")]
pub enum AuthRequest {
    #[default]
    Anonymous,
    CurrentTab {
        page_url: String,
        referer: String,
        user_agent: String,
        cookie_store_id: String,
        incognito: bool,
        cookies: Vec<CookieRecord>,
    },
    BraveProfile {
        profile_id: String,
    },
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct CookieRecord {
    pub name: String,
    pub value: String,
    pub domain: String,
    #[serde(default = "default_cookie_path")]
    pub path: String,
    #[serde(default)]
    pub secure: bool,
    #[serde(default)]
    pub http_only: bool,
    pub same_site: CookieSameSite,
    #[serde(default)]
    pub host_only: bool,
    pub expiration_date: Option<f64>,
    pub session: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CookieSameSite {
    NoRestriction,
    Lax,
    Strict,
    Unspecified,
}

fn default_cookie_path() -> String {
    "/".to_owned()
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum SelectionOption {
    #[serde(rename_all = "camelCase")]
    Preset {
        key: String,
        label: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        estimated_bytes: Option<u64>,
        #[serde(skip_serializing_if = "Option::is_none")]
        expected_container: Option<String>,
    },
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaSummary {
    pub extractor_key: String,
    pub media_id: String,
    pub webpage_url: String,
    pub title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thumbnail_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub uploader: Option<String>,
    pub is_live: bool,
    pub is_drm: bool,
    pub selections: Vec<SelectionOption>,
    pub probe_token: String,
    pub probed_at: u64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct StartDownloadRequest {
    pub job_id: String,
    pub probe_token: String,
    pub selection_key: String,
    pub auth: AuthRequest,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ClipMode {
    Fast,
    Exact,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct ClipSpec {
    pub start_ms: u64,
    pub end_ms: u64,
    pub mode: ClipMode,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct StartClipRequest {
    pub job_id: String,
    pub probe_token: String,
    pub selection_key: String,
    pub clip: ClipSpec,
    pub auth: AuthRequest,
    pub allow_full_download_fallback: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct CancelJobRequest {
    pub job_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(deny_unknown_fields)]
pub struct OutputActionRequest {
    pub output_token: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobQueued {
    pub job_id: String,
    pub position: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobProgress {
    pub job_id: String,
    pub stage: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub downloaded_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub percentage: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub speed_bytes_per_second: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eta_seconds: Option<f64>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FallbackRequired {
    pub job_id: String,
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub estimated_bytes: Option<u64>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobCompleted {
    pub job_id: String,
    pub output_token: String,
    pub final_path: String,
    pub filename: String,
    pub byte_size: u64,
    pub container: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub accuracy: Option<String>,
    pub extractor_key: String,
    pub media_id: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobState {
    pub job_id: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JobFailed {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub job_id: Option<String>,
    pub code: String,
    pub message: String,
    pub recoverable: bool,
}

impl JobFailed {
    pub fn from_error(job_id: Option<String>, error: &HostError) -> Self {
        let payload = ErrorPayload::from(error);
        Self {
            job_id,
            code: payload.code,
            message: payload.message,
            recoverable: payload.recoverable,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputActionResult {
    pub output_token: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn canonical_golden_envelopes_deserialize() {
        for fixture in [
            include_str!("../../protocol/fixtures/hello.json"),
            include_str!("../../protocol/fixtures/job-progress.json"),
            include_str!("../../protocol/fixtures/probe-result.json"),
        ] {
            let envelope: Envelope = serde_json::from_str(fixture).unwrap();
            assert_eq!(envelope.protocol_version, PROTOCOL_VERSION);
            assert!(!envelope.request_id.is_empty());
        }
        let hello: Envelope =
            serde_json::from_str(include_str!("../../protocol/fixtures/hello.json")).unwrap();
        let hello: HelloRequest = hello.parse_payload().unwrap();
        assert!(matches!(hello.browser_target, BrowserTarget::Brave));
        assert_eq!(hello.extension_version, "1.12.0");
    }

    #[test]
    fn auth_field_names_match_canonical_camel_case_contract() {
        let current_tab: AuthRequest = serde_json::from_value(json!({
            "mode": "current-tab",
            "pageUrl": "https://example.com/watch",
            "referer": "https://example.com/watch",
            "userAgent": "Brave fixture",
            "cookieStoreId": "1",
            "incognito": false,
            "cookies": [{
                "name": "session", "value": "secret", "domain": ".example.com",
                "path": "/", "secure": true, "httpOnly": true, "sameSite": "lax",
                "hostOnly": false, "session": true
            }]
        }))
        .unwrap();
        assert!(matches!(current_tab, AuthRequest::CurrentTab { .. }));

        let profile: AuthRequest = serde_json::from_value(json!({
            "mode": "brave-profile",
            "profileId": "Profile 2"
        }))
        .unwrap();
        assert!(
            matches!(profile, AuthRequest::BraveProfile { profile_id } if profile_id == "Profile 2")
        );
    }

    #[test]
    fn optional_response_fields_are_omitted_not_serialized_as_null() {
        let progress = JobProgress {
            job_id: "job".into(),
            stage: "planning".into(),
            detail: None,
            downloaded_bytes: None,
            total_bytes: None,
            percentage: None,
            speed_bytes_per_second: None,
            eta_seconds: None,
        };
        let value = serde_json::to_value(progress).unwrap();
        assert_eq!(value, json!({"jobId":"job","stage":"planning"}));
    }
}
